src/app/icd/ICDCheckInSender.cpp

```diff
@@ -0,0 +1,152 @@
+    VerifyOrDie(fabricTable != nullptr);
+    VerifyOrDie(symmetricKeyStore != nullptr);
+
+    mStorage           = storage;
+    mFabricTable       = fabricTable;
+    mSymmetricKeystore = symmetricKeyStore;
+
+    mAddressLookupHandle.SetListener(this);
+}
+
+ICDCheckInSender::ICDCheckInSender()
+{
+    mAddressLookupHandle.SetListener(this);
+}
+
+void ICDCheckInSender::Init(PersistentStorageDelegate * storage, FabricTable * fabricTable,
+                            Crypto::SymmetricKeystore * symmetricKeyStore)
+{
+    VerifyOrDie(storage != nullptr);
+    VerifyOrDie(fabricTable != nullptr);
+    VerifyOrDie(symmetricKeyStore != nullptr);
+
+    mStorage           = storage;
+    mFabricTable       = fabricTable;
+    mSymmetricKeystore = symmetricKeyStore;
+}
+
+void ICDCheckInSender::OnNodeAddressResolved(const PeerId & peerId, const AddressResolve::ResolveResult & result)
+{
+    mResolveInProgress = false;
+
+    ICDMonitoringEntry entry(mSymmetricKeystore);
+
+    const FabricInfo * f = mFabricTable->FindFabricWithCompressedId(peerId.GetCompressedFabricId());
+    ICDMonitoringTable table(*mStorage, f->GetFabricIndex(), CHIP_CONFIG_ICD_CLIENTS_SUPPORTED_PER_FABRIC, mSymmetricKeystore);
+    table.Find(peerId.GetNodeId(), entry);
+
+    if (CHIP_NO_ERROR != SendCheckInMsg(entry, result.address))
+    {
+        ChipLogError(AppServer, "Failed to send the ICD Check-In message");
+    }
+}
+
+void ICDCheckInSender::OnNodeAddressResolutionFailed(const PeerId & peerId, CHIP_ERROR reason)
+{
+    mResolveInProgress = false;
+    ChipLogProgress(AppServer, "Node Address resolution failed for ICD Check-In with Node ID " ChipLogFormatX64,
+                    ChipLogValueX64(peerId.GetNodeId()));
+}
+
+CHIP_ERROR ICDCheckInSender::SendCheckInMsg(ICDMonitoringEntry & entry, const Transport::PeerAddress & addr)
+{
+    uint8_t b[CheckinMessage::sMinPayloadSize] = { 0 };
+    MutableByteSpan output{ b };
+    CHIP_ERROR err;
+
+    // TODO retrieve Check-in counter
+    CounterType counter = 0;
+
+    // Prepare Check-in payload
+    err = CheckinMessage::GenerateCheckinMessagePayload(entry.key, counter, ByteSpan(), output);
+    if (err != CHIP_NO_ERROR)
+    {
+        return err;
+    }
+
+    if (InteractionModelEngine::GetInstance() == nullptr)
+    {
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    if (InteractionModelEngine::GetInstance()->GetExchangeManager() == nullptr)
+    {
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    if (InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager() == nullptr)
+    {
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    Messaging::ExchangeContext * exchangeContext = InteractionModelEngine::GetInstance()->GetExchangeManager()->NewContext(
+        InteractionModelEngine::GetInstance()
+            ->GetExchangeManager()
+            ->GetSessionManager()
+            ->CreateUnauthenticatedSession(addr, GetLocalMRPConfig().ValueOr(GetDefaultMRPConfig()))
```

Comment:
This is wrong.  CreateUnauthenticatedSession() takes the MRP config of the _remote_ as an argument...  But it doesn't matter too much, since we are not doing MRP anyway.  Still, either use the MRP config from the resolve, or document that the value here does not matter, but then you might as well just `GetDefaultMRPConfig()`.  Because the "local" thing is definitely 100% wrong here.

Reply:
thanks for the clarification. Will do.